ee/control/consumers/remoterestartconsumer/remoterestartconsumer.go

```diff
@@ -0,0 +1,125 @@
+package remoterestartconsumer
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"log/slog"
+	"time"
+
+	"github.com/kolide/launcher/ee/agent/types"
+)
+
+const (
+	// RemoteRestartActorType identifies this action/actor type, which performs
+	// a launcher restart when requested by the control server. This actor type
+	// belongs to the action subsystem.
+	RemoteRestartActorType = "remote_restart"
+
+	// restartDelay is the delay after receiving action before triggering the restart.
+	// We have a delay to allow the actionqueue.
+	restartDelay = 15 * time.Second
+)
+
+type RemoteRestartConsumer struct {
+	knapsack      types.Knapsack
+	slogger       *slog.Logger
+	signalRestart chan error
+	interrupt     chan struct{}
+	interrupted   bool
+}
+
+type remoteRestartAction struct {
+	RunID string `json:"run_id"` // the run ID for the launcher run to restart
+}
+
+func New(knapsack types.Knapsack) *RemoteRestartConsumer {
+	return &RemoteRestartConsumer{
+		knapsack:      knapsack,
+		slogger:       knapsack.Slogger().With("component", "remote_restart_consumer"),
+		signalRestart: make(chan error, 1),
+		interrupt:     make(chan struct{}, 1),
+	}
+}
+
+// Do implements the `actionqueue.actor` interface, and allows the actionqueue
+// to pass `remote_restart` type actions to this consumer. The actionqueue validates
+// that this action has not already been performed and that this action is still
+// valid (i.e. not expired). `Do` additionally validates that the `run_id` given in
+// the action matches the current launcher run ID.
+func (r *RemoteRestartConsumer) Do(data io.Reader) error {
+	var restartAction remoteRestartAction
+
+	if err := json.NewDecoder(data).Decode(&restartAction); err != nil {
+		return fmt.Errorf("decoding restart action: %w", err)
+	}
+
+	// The action's run ID indicates the current `runLauncher` that should be restarted.
+	// If the action's run ID does not match the current run ID, we assume the restart
+	// has already happened and does not need to happen again.
+	if restartAction.RunID == "" {
+		r.slogger.Log(context.TODO(), slog.LevelInfo,
+			"received remote restart action with blank launcher run ID -- discarding",
+		)
+		return nil
+	}
+	if restartAction.RunID != r.knapsack.GetRunID() {
+		r.slogger.Log(context.TODO(), slog.LevelInfo,
+			"received remote restart action for incorrect (assuming past) launcher run ID -- discarding",
+			"run_id", restartAction.RunID,
```

Comment:
should we namespace `run_id` for these as `action_run_id` or something? just wondering if this will get confusing or overwrite the `run_id` that we already include in logs

Reply:
Oh totally yup, let me update